Add `renew_cfd_transactions` API

luckysori opened the pull request:
As expected, renewing the CFD is the same as creating a CFD without a new lock transaction.

The tests are very repetitive, so I would be happy to start thinking about refactoring things there.

luckysori committed:
Take raw lock transaction in CommitTransaction constructor

We remove the `LockTransaction` type accordingly.

luckysori committed:
Rename tests file

luckysori committed:
Increase ContractExecutionTransaction::SIGNED_VBYTES constant

We were only testing against CETs with a single output, so the
`SIGNED_VBYTES` constant was based on a smaller transaction than
usual.

We could use a different constant for the case when the CET has a
single output, but it's just an edge case and the transaction fees are
very small.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -731,95 +745,55 @@ impl CommitTransaction {
+            taker_psbt.global.unsigned_tx.input,
+        ]
+        .concat(),
+        output: vec![vec![lock_output], maker_change, taker_change].concat(),
```
I think this double `Vec` is a bit confusing. Instead of `collect`ing all the individual iterators, you could concatenate them and use `iter::once` for the `lock_output`.

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -645,24 +645,38 @@ impl CommitTransaction {
+            let outpoint = OutPoint {
+                txid: lock_tx.txid(),
+                vout: vout as u32,
+            };
```
Perhaps a candidate for an extension trait?

thomaseizinger commented on cfd_protocol/src/lib.rs:
```diff
@@ -78,25 +71,112 @@ pub fn build_cfd_transactions(
+
+pub fn renew_cfd_transactions(
+    lock_tx: PartiallySignedTransaction,
+    (maker_pk, maker_lock_amount, maker_address, maker_punish_params): (
+        PublicKey,
+        Amount,
+        Address,
+        PunishParams,
+    ),
```
Why does `create_cfd_transactions` use `PartyParams` but `renew_cfd_transactions` doesn't?

luckysori replied:
I thought about it but it would be a bit redundant to pass in the lock PSBT twice.

thomaseizinger replied:
Didn't see that, fair enough :)

thomaseizinger reviewed (approved):
Nice work! Condensing down the tests sounds good to me. Should be useful for the integration later as well!

bonomat commented on cfd_protocol/src/lib.rs:
```diff
@@ -70,17 +70,16 @@ pub fn build_cfd_transactions(
 
     let commit_tx = CommitTransaction::new(
-        &lock_tx,
+        &lock_tx.global.unsigned_tx,
```
Should one prefer this over `lock_tx.extract_tx()`?

thomaseizinger replied:
`extract_tx` takes [`self`](https://docs.rs/bitcoin/0.27.0/bitcoin/util/psbt/struct.PartiallySignedTransaction.html#method.extract_tx) which consumes the entire PSBT. I guess in this case, @luckysori just wanted a reference to taking ownership of the PSBT itself was wasteful and would have probably resulted in a clone.

bonomat commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -372,6 +372,443 @@ fn create_cfd() {
 }
 
+#[test]
+fn renew_cfd() {
```
I guess this asks for some refactoring and code-reuse between the two different tests :)

bonomat reviewed (approved):
Look good, nice work 👋👋

luckysori committed:
Use iter::chain and iter::once

Instead of `vec![vec![]].concat()`.

luckysori committed:
Define and implement TransactionExt::outpoint

luckysori committed:
Add create test helper

luckysori committed:
Rename finalize_spend_transaction arguments

To convey that the order in which these pairs of arguments are passed
in is irrelevant.

luckysori committed:
Refactor cfd tests

In order to reuse code and to make it easier to add new tests (e.g.
negative tests) in the future.

luckysori committed:
Add renew_cfd_transactions API

thomaseizinger commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -46,290 +47,77 @@ fn create_cfd() {
+            rev_pk: taker_rev_pk,
+            pub_sk: taker_pub_sk,
+            pub_pk: taker_pub_pk,
+        },
```
Just call this `taker` and refer to fields as `taker.sk`.

luckysori committed:
Do not destructure CfdKeys type
